### PR TITLE
Initialize a fresh detector for each context

### PR DESCRIPTION
**Pull request by mvz**
The existing implementation used the same detector for all contexts, which resulted in constantly passing around the current context and other values as parameters. This in turn has lead to a large number of
cases of feature envy and similar code smells.

By using each detector instance for just one context, we can migrate to a design where more of the state of the ongoing detection is stored in the detector and less is passed around.

### CODE REVIEW

**Comment by mvz** [comment]
I think this would be a real improvement to Reek's design. WDYT?

File: lib/reek/detector_repository.rb [hunk 1]
```diff
@@ -35,31 +35,25 @@ def initialize(smell_types: self.class.smell_types,
                    configuration: {})
       @configuration = configuration
       @smell_types   = smell_types
-      @detectors     = smell_types.map { |klass| klass.new configuration_for(klass) }
     end
 
     def examine(context)
-      smell_detectors_for(context.type).flat_map do |detector|
-        detector.run_for(context)
+      smell_detectors_for(context.type).flat_map do |klass|
```
Comment (Drenmi) [inline lib/reek/detector_repository.rb]: At first glance, it doesn't look like `#smell_detectors_for` would return a two-dimensional array. Could this get away with just `#map`?

File: lib/reek/detector_repository.rb [hunk 2]
```diff
@@ -35,31 +35,25 @@ def initialize(smell_types: self.class.smell_types,
...
 
     def smell_detectors_for(type)
-      enabled_detectors.select do |detector|
-        detector.contexts.include? type
-      end
-    end
-
-    def enabled_detectors
-      detectors.select { |detector| detector.config.enabled? }
+      smell_types.select { |detector| detector.contexts.include? type }
```
Comment (Drenmi) [inline lib/reek/detector_repository.rb]: There's a cop (which is currently disabled by default) `Style/MethodCallWithArgsParentheses` which checks for the use of parens around method calls with arguments.

It is pretty reasonable if run with `IgnoreMacros: true` (default) and you can selectively whitelist methods as well. I am pushing to make it enabled in the next release of RuboCop, but if you're brave you can enable it now. 🙂

File: lib/reek/detector_repository.rb (hunk 1 above)
Comment (mvz) [inline lib/reek/detector_repository.rb]: It doesn't but the block returns an array.

Comment (Drenmi) [inline lib/reek/detector_repository.rb]: Right. 🙂

File: spec/reek/smell_detectors/irresponsible_module_spec.rb [hunk 3]
```diff
@@ -49,17 +49,6 @@ class Charlie
...
-    it "does not report re-opened #{scope} in the same file" do
-      src = <<-EOS
-        # This comment describes Alfa
-        #{scope} Alfa; end
-
-        #{scope} Alfa; def bravo; end; end
-      EOS
-
-      expect(src).not_to reek_of(:IrresponsibleModule)
-    end
```
Comment (troessner) [inline spec/reek/smell_detectors/irresponsible_module_spec.rb]: How is this change relevant to this PR?

**Comment by troessner** [comment]
One tiny comment and then we're good to roll. I love this change overall, the "old" way always bothered me.

File: spec/reek/smell_detectors/irresponsible_module_spec.rb (hunk 3 above)
Comment (mvz) [inline spec/reek/smell_detectors/irresponsible_module_spec.rb]: This change makes it impossible for detectors to keep state between sniffing one context and the next. Therefore, the behavior specified here can no longer work. I don't think this is a great loss, since:

* Why would you open a class or module twice in a single file, *and* define methods in it both times?
* None of the other detectors keep state for the same kind of case, e.g., TooManyMethods, and UnusedPrivateMethod.

File: lib/reek/detector_repository.rb (hunk 1 above)
Comment (mvz) [inline lib/reek/detector_repository.rb]: 🙂

File: spec/reek/smell_detectors/irresponsible_module_spec.rb (hunk 3 above)
Comment (troessner) [inline spec/reek/smell_detectors/irresponsible_module_spec.rb]: Gotcha, makes sense :)

**Comment by mvz** [comment]
Awesomeness!
